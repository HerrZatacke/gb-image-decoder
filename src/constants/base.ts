import { RGBNPalette } from '../Types';
import { BlendMode } from './blendModes';


export const BLACK = 'FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF';
export const WHITE = '00000000000000000000000000000000';

export const SKIP_LINE = 'skip';

export const TILE_PIXEL_WIDTH = 8;

export const TILE_PIXEL_HEIGHT = 8;

export const TILES_PER_LINE = 20;

export const TILES_PER_COLUMN = 18;

export const FRAME_SIZE = 2;

export const FRAME_TILES = FRAME_SIZE * 2;

export const DEFAULT_FULL_PIXEL_HEIGHT = TILES_PER_COLUMN * TILE_PIXEL_HEIGHT;

export const DEFAULT_FULL_PIXEL_WIDTH = TILES_PER_LINE * TILE_PIXEL_WIDTH;

export const BW_PALETTE: number[] = [0xffffff, 0xaaaaaa, 0x555555, 0x000000];

export const BW_PALETTE_HEX: string[] = ['#ffffff', '#aaaaaa', '#555555', '#000000'];

export const BLACK_LINE: string[] = Array(TILES_PER_LINE).fill(BLACK);

export const WHITE_LINE: string[] = Array(TILES_PER_LINE).fill(WHITE);

export const defaultPalette: RGBNPalette = {
  r: [0, 84, 172, 255],
  g: [0, 84, 172, 255],
  b: [0, 84, 172, 255],
  n: [0, 85, 170, 255],
  blend: BlendMode.MULTIPLY,
};
