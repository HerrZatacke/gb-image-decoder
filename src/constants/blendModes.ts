export enum BlendMode {
  NORMAL = 'normal', // skip N-Layer
  NORMAL_S = 'normal_s', // This is the default setting and draws new shapes on top of the existing canvas content.
  LIGHTEN = 'lighten', // Retains the lightest pixels of both layers.
  SCREEN = 'screen', // The pixels are inverted, multiplied, and inverted again. A lighter picture is the result (opposite of multiply)
  DODGE = 'dodge', // Divides the bottom layer by the inverted top layer.
  ADDITION = 'addition', // Where both shapes overlap, the color is determined by adding color values.
  DARKEN = 'darken', // Retains the darkest pixels of both layers.
  MULTIPLY = 'multiply', // The pixels of the top layer are multiplied with the corresponding pixels of the bottom layer. A darker picture is the result.
  BURN = 'burn', // Divides the inverted bottom layer by the top layer, and then inverts the result.
  OVERLAY = 'overlay', // A combination of multiply and screen. Dark parts on the base layer become darker, and light parts become lighter.
  SOFTLIGHT = 'softlight', // A softer version of hard-light. Pure black or white does not result in pure black or white.
  HARDLIGHT = 'hardlight', // Like overlay, a combination of multiply and screen — but instead with the top layer and bottom layer swapped.
  DIFFERENCE = 'difference', // Subtracts the bottom layer from the top layer — or the other way round — to always get a positive value.
  EXCLUSION = 'exclusion', // Like difference, but with lower contrast.
}

export const blendModeNewName = (oldName?: BlendMode): GlobalCompositeOperation => {
  switch (oldName) {
    case BlendMode.NORMAL:
      return 'source-over';
    case BlendMode.NORMAL_S:
      return 'source-over';
    case BlendMode.LIGHTEN:
      return 'lighten';
    case BlendMode.SCREEN:
      return 'screen';
    case BlendMode.DODGE:
      return 'color-dodge';
    case BlendMode.ADDITION:
      return 'lighter';
    case BlendMode.DARKEN:
      return 'darken';
    case BlendMode.MULTIPLY:
      return 'multiply';
    case BlendMode.BURN:
      return 'color-burn';
    case BlendMode.OVERLAY:
      return 'overlay';
    case BlendMode.SOFTLIGHT:
      return 'soft-light';
    case BlendMode.HARDLIGHT:
      return 'hard-light';
    case BlendMode.DIFFERENCE:
      return 'difference';
    case BlendMode.EXCLUSION:
      return 'exclusion';
    default:
      return 'multiply';
  }
};
