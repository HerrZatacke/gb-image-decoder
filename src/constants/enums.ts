export enum ChannelKey {
  R = 'r',
  G = 'g',
  B = 'b',
  N = 'n',
}

export enum ExportFrameMode {
  FRAMEMODE_KEEP = 'keep',
  FRAMEMODE_CROP = 'crop',
  FRAMEMODE_SQUARE_BLACK = 'square_black',
  FRAMEMODE_SQUARE_WHITE = 'square_white',
}

export enum Rotation {
  DEG_0 = 0,
  DEG_90 = 1,
  DEG_180 = 2,
  DEG_270 = 3,
}

export const channels = [ChannelKey.R, ChannelKey.G, ChannelKey.B, ChannelKey.N];
