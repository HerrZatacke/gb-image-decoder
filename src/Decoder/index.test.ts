/* eslint-disable max-len */
import { beforeEach, describe, expect, test } from 'vitest';
import { hash } from 'ohash';
import { Decoder } from '.';
import { DecoderOptions, DecoderUpdateParams, ExportFrameMode } from '../index';
import tiles1x1 from '../../test/data/tiles/monochrome/1x1';
import tiles16x14 from '../../test/data/tiles/monochrome/16x14';
import tiles20x18 from '../../test/data/tiles/monochrome/20x18';
import tiles20x86 from '../../test/data/tiles/monochrome/20x86';
import { bw, red } from '../../test/data/palettes';
import { TILES_PER_LINE } from '../constants/base';
import { writeImageFileFromImageData } from '../../test/helpers/writeImageFileFromImageData';
import { getCanvasImageData } from '../../test/helpers/ getCanvasImageData';

const TEST_SCALE_FACTOR = 2;

const tileSets: Record<string, string[]> = {
  '1x1': tiles1x1,
  '16x14': tiles16x14,
  '20x18': tiles20x18,
  '20x86': tiles20x86,
};

describe('Monochrome Decoder', () => {
  let defaultDecoderOptions: DecoderOptions = {};
  let canvas: HTMLCanvasElement;

  beforeEach(() => {
    canvas = document.createElement('canvas');
    defaultDecoderOptions = {
      tilesPerLine: TILES_PER_LINE,
    };
  });

  describe.each([
    { tileKey: '1x1', tilesPerLine: 1, imageStartLine: 0 },
    { tileKey: '16x14', tilesPerLine: 16, imageStartLine: 2 },
    { tileKey: '20x18', tilesPerLine: 20, imageStartLine: 2 },
    { tileKey: '20x86', tilesPerLine: 20, imageStartLine: 13 },
  ])('using $tileKey raw data', async ({ tileKey, tilesPerLine, imageStartLine }) => {
    const constructDecoderOptions = {
      ...defaultDecoderOptions,
      tilesPerLine,
    };

    const decoderUpdateParams: Omit<DecoderUpdateParams, 'canvas'> = {
      framePalette: red,
      palette: bw,
      tiles: tileSets[tileKey],
      imageStartLine,
    };

    test('renders on to a given canvas', async () => {
      const decoder = new Decoder(constructDecoderOptions);

      decoder.update({
        canvas,
        ...decoderUpdateParams,
      });

      await writeImageFileFromImageData(`mono_${tileKey}_default.png`, canvas);

      expect(hash(getCanvasImageData(canvas).data)).toMatchSnapshot();
    });

    test.each([
      { frameMode: ExportFrameMode.FRAMEMODE_KEEP },
      { frameMode: ExportFrameMode.FRAMEMODE_CROP },
      { frameMode: ExportFrameMode.FRAMEMODE_SQUARE_BLACK },
      { frameMode: ExportFrameMode.FRAMEMODE_SQUARE_WHITE },
    ])('renders a scaled canvas with framemode: $frameMode', async ({ frameMode }) => {
      const decoder = new Decoder(constructDecoderOptions);

      decoder.update({
        canvas: null,
        ...decoderUpdateParams,
      });

      const scaledCanvas = decoder.getScaledCanvas(TEST_SCALE_FACTOR, frameMode);

      await writeImageFileFromImageData(`mono_${tileKey}_scaled-${frameMode}.png`, scaledCanvas);
      expect(hash(getCanvasImageData(scaledCanvas).data)).toMatchSnapshot();
    });
  });

});
