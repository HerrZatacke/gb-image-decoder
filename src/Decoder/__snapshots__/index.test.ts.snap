// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Monochrome Decoder > using '1x1' raw data > renders a scaled canvas with framemode: 'crop' 1`] = `"jwkXkBq86d"`;

exports[`Monochrome Decoder > using '1x1' raw data > renders a scaled canvas with framemode: 'keep' 1`] = `"jwkXkBq86d"`;

exports[`Monochrome Decoder > using '1x1' raw data > renders a scaled canvas with framemode: 'square_black' 1`] = `"jwkXkBq86d"`;

exports[`Monochrome Decoder > using '1x1' raw data > renders a scaled canvas with framemode: 'square_white' 1`] = `"jwkXkBq86d"`;

exports[`Monochrome Decoder > using '1x1' raw data > renders on to a given canvas 1`] = `"HMPDkCwruk"`;

exports[`Monochrome Decoder > using '16x14' raw data > renders a scaled canvas with framemode: 'crop' 1`] = `"vVfl7TJX14"`;

exports[`Monochrome Decoder > using '16x14' raw data > renders a scaled canvas with framemode: 'keep' 1`] = `"vVfl7TJX14"`;

exports[`Monochrome Decoder > using '16x14' raw data > renders a scaled canvas with framemode: 'square_black' 1`] = `"vVfl7TJX14"`;

exports[`Monochrome Decoder > using '16x14' raw data > renders a scaled canvas with framemode: 'square_white' 1`] = `"vVfl7TJX14"`;

exports[`Monochrome Decoder > using '16x14' raw data > renders on to a given canvas 1`] = `"oPjwTsSVGy"`;

exports[`Monochrome Decoder > using '20x18' raw data > renders a scaled canvas with framemode: 'crop' 1`] = `"vVfl7TJX14"`;

exports[`Monochrome Decoder > using '20x18' raw data > renders a scaled canvas with framemode: 'keep' 1`] = `"IUQ1y9uAd6"`;

exports[`Monochrome Decoder > using '20x18' raw data > renders a scaled canvas with framemode: 'square_black' 1`] = `"d5Zw21fWDW"`;

exports[`Monochrome Decoder > using '20x18' raw data > renders a scaled canvas with framemode: 'square_white' 1`] = `"831JE9T3f2"`;

exports[`Monochrome Decoder > using '20x18' raw data > renders on to a given canvas 1`] = `"zHEfqPiH13"`;

exports[`Monochrome Decoder > using '20x86' raw data > renders a scaled canvas with framemode: 'crop' 1`] = `"vVfl7TJX14"`;

exports[`Monochrome Decoder > using '20x86' raw data > renders a scaled canvas with framemode: 'keep' 1`] = `"rSAwi2jhQF"`;

exports[`Monochrome Decoder > using '20x86' raw data > renders a scaled canvas with framemode: 'square_black' 1`] = `"NEe0NAJBDK"`;

exports[`Monochrome Decoder > using '20x86' raw data > renders a scaled canvas with framemode: 'square_white' 1`] = `"vEwjOlU8H6"`;

exports[`Monochrome Decoder > using '20x86' raw data > renders on to a given canvas 1`] = `"NobxBYWa3H"`;
